src/main/java/org/manifold/compiler/front/FunctionInvocationVertex.java

```diff
@@ -107,9 +113,74 @@ private void elaborateNodeInstantiation(Value function,
   }
+
+  private void elaborateNonPrimitiveFunction(Value f,
+      ExpressionVertex input) throws Exception {
+    log.debug("function invocation is non-primitive elaboration");
+    FunctionValue function = (FunctionValue) f;
+    FunctionTypeValue signature = (FunctionTypeValue) f.getType();
+    TupleTypeValue inputType = (TupleTypeValue) signature.getInputType();
+    TupleTypeValue outputType = (TupleTypeValue) signature.getOutputType();
+    // calculate renaming map (body -> this.exprGraph)
+    Map<VariableReferenceVertex, VariableReferenceVertex> renamingMap =
+        new HashMap<>();
+    for (Entry<VariableIdentifier, VariableReferenceVertex> entry
+        : function.getBody().getVariableVertices().entrySet()) {
+      VariableReferenceVertex varTarget;
+      if (getExpressionGraph().containsVariable(entry.getKey())) {
+        // if the main graph contains this variable, don't rename; use existing
+        log.debug("preserving variable " + entry.getKey());
+        varTarget = getExpressionGraph().getVariableVertex(entry.getKey());
+      } else {
+        // otherwise, create a fresh variable in the main graph and map to that
+        String oldName = entry.getKey().getName();
+        Random rng = new Random();
+        while (true) {
+          // choose a random suffix and mangle
+          Integer i = rng.nextInt(Integer.MAX_VALUE);
+          String newName = oldName + "_" + i.toString();
+          List<String> newNames = new ArrayList<>();
+          newNames.add(newName);
+          VariableIdentifier newID = new VariableIdentifier(newNames);
+          if (!getExpressionGraph().containsVariable(newID)) {
+            log.debug("renaming local variable " + entry.getKey()
+                + " to " + newID);
+            getExpressionGraph().addVertex(newID);
+            varTarget = getExpressionGraph().getVariableVertex(newID);
+            break;
+          }
+        }
+      }
+      renamingMap.put(entry.getValue(), varTarget);
+    }
+    // identify main graph input and output edges
+    // we assume there's exactly one of each
+    ExpressionEdge mainGraphInput = inputEdge;
+    log.debug("main graph input is " + mainGraphInput);
+    ExpressionEdge mainGraphOutput = null;
+    for (ExpressionEdge e : getExpressionGraph().getEdgesFromSource(this)) {
+      mainGraphOutput = e;
+      break;
+    }
+    log.debug("main graph output is " + mainGraphOutput);
+    // identify subgraph (body) input and output vertices
+    ExpressionVertex subGraphInput = function.getInputVertex();
+    ExpressionVertex subGraphOutput = function.getOutputVertex();
+    // perform copy
+    getExpressionGraph().addSubExpressionGraph(function.getBody(),
+        mainGraphInput, subGraphInput, mainGraphOutput, subGraphOutput,
+        renamingMap);
+    // destroy the function edge
+    getExpressionGraph().removeEdge(functionEdge);
+    // now remove this vertex from the graph
+    getExpressionGraph().removeVertex(this);
```

Comment:
I think this is redundant since in addSubExpressionGraph:

```
    // first, remove the function invocation vertex & the edges dealing with it
    this.allVertices.remove(mainGraphInput.getTarget());
    this.edges.removeAll(ImmutableList.of(mainGraphInput, mainGraphOutput));
```